<div align="center"> 
    
# ***Manufacturing Downtime Optimization***
##### Data-Driven Manufacturing Performance, Downtime & Predictive Analytics

 </div>

--- 

## 🔍 Project Overview
##### This project focuses on analyzing manufacturing downtime, production performance, machine efficiency, quality, and energy consumption to identify operational inefficiencies and their impact on overall manufacturing performance.

##### It aims to identify the main drivers of downtime and production losses, evaluate machine and shift performance, monitor OEE and quality indicators, and generate data-driven insights to support operational improvement and downtime reduction.

##### The project is designed for manufacturing operations and production teams seeking to improve equipment utilization, reduce production losses, optimize operational efficiency, and support data-driven decision-making through analytics and predictive insights .


---

## 🛠️ Technology Stack 

| **Category** | **Technology** |
|:---------|:-----------|
| ***Programming Language*** | *Python* |
| ***Data Manipulation*** | *Pandas* , *Numpy* |
| ***Data Visualization*** | *Matplotlib, Seaborn , Power BI*  |
| ***Database Language*** | *SQL* |
| ***Database Management System*** | *SQL Server* |
| ***Database Design*** | *Star Schema* |
| ***Data Source*** | *manufacturing Dataset* |

--- 


<div style="color: Orange; text-align: center; padding: 7px; border-radius: 5px; font-weight: bold; font-size: 25px;">Analysis Implementation 
</div> 

### **1- Import Libraries**
##### Import the required Python libraries for data manipulation, visualization, and analysis.

In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 

### **2- Data Acquisition**
##### Load the manufacturing.csv dataset as the source data for the analysis .

In [2]:
try :
    Source_File = 'manufacturing.xlsx'
    Sheet_Name = 'manufacturing_Sheet'
    manufacturing = pd.read_excel(Source_File , sheet_name = Sheet_Name )
except FileNotFoundError :
    print("File missing")
else :
    print("File loaded successfully")
finally :
    print("Execution finished")

File loaded successfully
Execution finished


### **3- Dataset Inspection**
#### Inspect the structure, dimensions, data types, and overall quality of the datasets.
--- 

In [3]:
# Preview the first 5 records to understand the dataset structure.
manufacturing.head()

,date,shift,machine_id,product_id,operator_id,planned_time_min,downtime_min,downtime_reason,operating_time_min,total_units,...,has_breakdown,has_defect,low_OEE,OEE_category,shift_enc,machine_id_enc,product_id_enc,operator_id_enc,downtime_reason_enc,defect_type_enc
0,2025-01-01,Shift_1,M01,P_E,OP026,480,47,No Downtime,433,511,...,0,1,0,Fair(65-75%),0,0,4,25,3,4
1,2025-01-01,Shift_2,M01,P_C,OP027,480,110,No Downtime,370,368,...,0,0,0,Fair(65-75%),1,0,2,26,3,2
2,2025-01-01,Shift_3,M01,P_D,OP006,480,6,Setup,474,483,...,0,0,0,Fair(65-75%),2,0,3,5,4,2
3,2025-01-01,Shift_1,M02,P_E,OP010,480,78,No Downtime,402,495,...,0,0,1,Poor(50-65%),0,1,4,9,3,2
4,2025-01-01,Shift_2,M02,P_E,OP029,480,93,Material Shortage,387,579,...,0,0,1,Poor(50-65%),1,1,4,28,2,2


In [4]:
# Preview the last 5 records to verify the dataset ends correctly.
manufacturing.tail() 

,date,shift,machine_id,product_id,operator_id,planned_time_min,downtime_min,downtime_reason,operating_time_min,total_units,...,has_breakdown,has_defect,low_OEE,OEE_category,shift_enc,machine_id_enc,product_id_enc,operator_id_enc,downtime_reason_enc,defect_type_enc
54745,2025-12-31,Shift_2,M49,P_D,OP032,480,117,No Downtime,363,646,...,0,1,1,Poor(50-65%),1,48,3,31,3,4
54746,2025-12-31,Shift_3,M49,P_A,OP028,480,28,No Downtime,452,810,...,0,1,0,Fair(65-75%),2,48,0,27,3,4
54747,2025-12-31,Shift_1,M50,P_D,OP002,480,29,No Downtime,451,450,...,0,1,0,Fair(65-75%),0,49,3,1,3,0
54748,2025-12-31,Shift_2,M50,P_C,OP032,480,26,No Downtime,454,682,...,0,0,0,Good(75-85%),1,49,2,31,3,2
54749,2025-12-31,Shift_3,M50,P_A,OP011,480,29,Setup,451,538,...,0,0,0,Good(75-85%),2,49,0,10,4,2


In [5]:
# Check the number of rows and columns in the dataset.
manufacturing.shape

(54750, 38)

In [6]:
manufacturing.dtypes

date                     datetime64[us]
shift                               str
machine_id                          str
product_id                          str
operator_id                         str
planned_time_min                  int64
downtime_min                      int64
downtime_reason                     str
operating_time_min                int64
total_units                       int64
good_units                        int64
reject_units                      int64
defect_type                         str
energy_kwh                      float64
availability                    float64
performance                     float64
quality                         float64
OEE                             float64
month                             int64
day_of_week                       int64
is_weekend                        int64
week                              int64
downtime_rate                   float64
reject_rate                     float64
energy_per_unit                 float64


In [7]:
# Inspect data types and non-null values for each column.
manufacturing.info()

<class 'pandas.DataFrame'>
RangeIndex: 54750 entries, 0 to 54749
Data columns (total 38 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   date                   54750 non-null  datetime64[us]
 1   shift                  54750 non-null  str           
 2   machine_id             54750 non-null  str           
 3   product_id             54750 non-null  str           
 4   operator_id            54750 non-null  str           
 5   planned_time_min       54750 non-null  int64         
 6   downtime_min           54750 non-null  int64         
 7   downtime_reason        54750 non-null  str           
 8   operating_time_min     54750 non-null  int64         
 9   total_units            54750 non-null  int64         
 10  good_units             54750 non-null  int64         
 11  reject_units           54750 non-null  int64         
 12  defect_type            54750 non-null  str           
 13  energy_kwh  

#### `Observation`
- The dataset contains 54750 records and 38 features, providing sufficient data for business analysis.
- The dataset includes numerical, categorical, and datetime features.
- Data quality is generally good, with no missing values.

In [8]:
# Display all column names for inspection . 
manufacturing.columns

Index(['date', 'shift', 'machine_id', 'product_id', 'operator_id',
       'planned_time_min', 'downtime_min', 'downtime_reason',
       'operating_time_min', 'total_units', 'good_units', 'reject_units',
       'defect_type', 'energy_kwh', 'availability', 'performance', 'quality',
       'OEE', 'month', 'day_of_week', 'is_weekend', 'week', 'downtime_rate',
       'reject_rate', 'energy_per_unit', 'oee_gap', 'production_efficiency',
       'no_downtime', 'has_breakdown', 'has_defect', 'low_OEE', 'OEE_category',
       'shift_enc', 'machine_id_enc', 'product_id_enc', 'operator_id_enc',
       'downtime_reason_enc', 'defect_type_enc'],
      dtype='str')

In [9]:
# Generate summary statistics for numerical features.
manufacturing.describe()  

,date,planned_time_min,downtime_min,operating_time_min,total_units,good_units,reject_units,energy_kwh,availability,performance,...,no_downtime,has_breakdown,has_defect,low_OEE,shift_enc,machine_id_enc,product_id_enc,operator_id_enc,downtime_reason_enc,defect_type_enc
count,54750,54750.0,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,...,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000,54750.000000
mean,2025-07-02 00:00:00,480.0,59.337114,420.662886,597.696968,564.384840,33.312128,234.969597,0.876381,0.849866,...,0.008785,0.098247,0.150429,0.309443,1.000000,24.500000,1.999836,19.504731,2.507068,2.010338
min,2025-01-01 00:00:00,480.0,0.000000,361.000000,300.000000,270.000000,4.000000,120.000000,0.752000,0.700000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2025-04-02 00:00:00,480.0,30.000000,391.000000,447.000000,422.000000,18.000000,177.360000,0.815000,0.775000,...,0.000000,0.000000,0.000000,0.000000,0.000000,12.000000,1.000000,10.000000,3.000000,2.000000
50%,2025-07-02 00:00:00,480.0,59.000000,421.000000,597.000000,564.000000,30.000000,234.690000,0.877000,0.850000,...,0.000000,0.000000,0.000000,0.000000,1.000000,24.500000,2.000000,19.000000,3.000000,2.000000
75%,2025-10-01 00:00:00,480.0,89.000000,450.000000,747.000000,705.000000,46.000000,292.837500,0.938000,0.925000,...,0.000000,0.000000,0.000000,1.000000,2.000000,37.000000,3.000000,29.000000,3.000000,2.000000
max,2025-12-31 00:00:00,480.0,119.000000,480.000000,899.000000,889.000000,90.000000,350.000000,1.000000,1.000000,...,1.000000,1.000000,1.000000,1.000000,2.000000,49.000000,4.000000,39.000000,4.000000,4.000000
std,NaN,0.0,34.634165,34.634165,172.958778,164.325186,18.736742,66.586835,0.072192,0.086556,...,0.093319,0.297650,0.357495,0.462268,0.816504,14.431001,1.417870,11.524369,1.067562,0.575837


#### `Observation`
- There are 3 shifts, with each shift lasting 8 hours (480 minutes).
- The average machine downtime is approximately 59 minutes per shift, while the maximum recorded downtime is 119 minutes, which is nearly 2 hours.
- The average actual machine operating time is approximately 420.66 minutes, equivalent to around 7 hours of effective operating time per shift.
- The average production volume is close to the median, indicating a relatively balanced distribution of units produced.
- Energy consumption ranges from 120 to 350 kWh, with an average consumption of approximately 235 kWh per shift.
- The average Overall Equipment Effectiveness (OEE) is approximately 70%, indicating room for improvement in overall equipment performance.

In [11]:
# Generate summary statistics for categorical features.
manufacturing.describe(include=['object','string']) 

,shift,machine_id,product_id,operator_id,downtime_reason,defect_type,OEE_category
count,54750,54750,54750,54750,54750,54750,54750
unique,3,50,5,40,5,5,5
top,Shift_1,M01,P_A,OP010,No Downtime,No Defect,Fair(65-75%)
freq,18250,1095,11032,1455,38345,46514,20654


#### `Observation`
- The dataset contains 3 types of shifts, with Shift 1 being the most frequent.
- There are 5 categories for overall machine efficiency, with the most common being the Fair (65-75%) tier.
- The dataset is perfectly balanced across all assets, where each of the 50 machines (M01 to M50) ran for an average of 1,095 shifts throughout the year, operated by a total of 40 operators (OP01 to OP40).

<div style="color: Green;  padding: 7px; border-radius: 5px; font-weight: bold; font-size: 25px;"> Data Cleaning </div>

---

In [9]:
# Display Columns with Missing Values (> 0%)
missing_percentage = (manufacturing.isnull().mean().mul(100).round(2).sort_values(ascending=False))
missing_percentage = missing_percentage[missing_percentage > 0]
print(missing_percentage) 

Series([], dtype: float64)


#### `Observation`
-  The missing values represent  0% of the dataset , So the manufacturing dataset hasn't contain missing values  .

In [10]:
# Identify duplicate records in the dataset.
manufacturing.duplicated().sum() 

np.int64(0)

#### `Observation`
- the manufacturing dataset hasn't contain duplicate values 

In [25]:
# Count the number of unique values in each column.
manufacturing.nunique() 

date                       365
shift                        3
machine_id                  50
product_id                   5
operator_id                 40
planned_time_min             1
downtime_min               120
downtime_reason              5
operating_time_min         120
total_units                600
good_units                 619
reject_units                87
defect_type                  5
energy_kwh               20931
availability               120
performance                301
quality                     94
OEE                        499
month                       12
day_of_week                  7
is_weekend                   2
week                        52
downtime_rate              120
reject_rate                928
energy_per_unit           8800
oee_gap                    499
production_efficiency      928
no_downtime                  2
has_breakdown                2
has_defect                   2
low_OEE                      2
OEE_category                 5
shift_en

In [27]:
# Standardize column names by:Removing leading/trailing whitespace,Converting names to Title ,Replacing spaces and hyphens with underscores.
# snake_case 
manufacturing.columns = (manufacturing.columns.str.strip().str.replace(" " ,"_" , regex = False).str.replace("/" ,"_" , regex = False)
    .str.replace("-" ,"_" , regex = False).str.lower()) 
print(manufacturing.columns)

Index(['date', 'shift', 'machine_id', 'product_id', 'operator_id',
       'planned_time_min', 'downtime_min', 'downtime_reason',
       'operating_time_min', 'total_units', 'good_units', 'reject_units',
       'defect_type', 'energy_kwh', 'availability', 'performance', 'quality',
       'oee', 'month', 'day_of_week', 'is_weekend', 'week', 'downtime_rate',
       'reject_rate', 'energy_per_unit', 'oee_gap', 'production_efficiency',
       'no_downtime', 'has_breakdown', 'has_defect', 'low_oee', 'oee_category',
       'shift_enc', 'machine_id_enc', 'product_id_enc', 'operator_id_enc',
       'downtime_reason_enc', 'defect_type_enc'],
      dtype='str')


In [19]:
# Replaced shift values with more understandable labels: Morning, Evening, and Night shifts.
manufacturing['shift'] = manufacturing['shift'].replace({'Shift_1':'Morning' ,'Shift_2': 'Evening' ,'Shift_3': 'Night'})

In [20]:
# Dropped the production_efficiency column because it represents the same values as the quality column.
manufacturing.drop(columns='production_efficiency' , inplace = True )

<div style="color: Green;  padding: 7px; border-radius: 5px; font-weight: bold; font-size: 25px;"> Data Warehouse Design </div>

In [ ]:
# Building the dims , fact Tables


<div style="color: Green;  padding: 7px; border-radius: 5px; font-weight: bold; font-size: 25px;"> Exploratory Data Analysis </div>